source/Octopus.Cli/Util/RepositoryExtensions.cs

```diff
@@ -0,0 +1,145 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Octopus.Cli.Infrastructure;
+using Octopus.Client.Exceptions;
+using Octopus.Client.Extensibility;
+using Octopus.Client.Logging;
+using Octopus.Client.Model;
+using Octopus.Client.Repositories.Async;
+
+namespace Octopus.Cli.Util
+{
+    public static class RepositoryExtensions
+    {
+        private static readonly ILog Logger = LogProvider.GetLogger(typeof(RepositoryExtensions));
+
+        private static async Task<TResource> FindByNameOrIdOrFail<T, TResource>(this T repository,
+            Func<string, Task<TResource>> findByNameFunc, string fixedIdPrefix, string typeDescription, string nameOrId,
+            string inDescription = "", bool skipLog = false)
+            where T : IGet<TResource> where TResource : Resource, INamedResource
+        {
+            TResource resourceById;
+            if (!Regex.IsMatch(nameOrId, $@"^{Regex.Escape(fixedIdPrefix)}-\d+$",
+                RegexOptions.Compiled | RegexOptions.IgnoreCase))
+            {
+                resourceById = null;
+            }
+            else
+            {
+                try
+                {
+                    resourceById = await repository.Get(nameOrId)
+                        .ConfigureAwait(false);
+                }
+                catch (OctopusResourceNotFoundException)
+                {
+                    resourceById = null;
+                }
+            }
+
+            var resourceByName = await findByNameFunc(nameOrId)
```

Comment:
Can this ever throw? (ie, a notfoundexception or something)

Reply:
Not under conventional circumstances, but there's no harm in catching `OctopusResourceNotFoundException` for consistency with the by-Id Get just above. So I've done that now.